### DISCUSSION TITLE
Loading Custom Settings with Factory Reset Command

### DISCUSSION THREAD

**Discussion by NsinghP** [arendst/Tasmota#11608] [Q&A] [answered]
Hello, 
I have made few  4 Channel, ESP8266-12F based devices which I am using  at home which are running Tasmota. Operation has been excellent. Now I have configuration question as follows:

(a)	I have made and activated a template named “MYHOME’ with following mappings: 

GPIO-14 --->Switch-1, GPIO-13 --->Switch-2, GPIO-12 --->Switch-3, GPIO-3 --->Switch-4 
GPIO-0 --->Relay_i-1, GPIO-4 --->Relay_i-2, GPIO-5 --->Relay_i-3, GPIO-1 --->Relay_i-4

(b)	Also on “Configure Other” page of Tasmota, I have following custom settings:

Template: {"NAME":"MYHOME","GPIO":[258,259,1,163,256,257,1,1,162,161,160,1,1,1],"FLAG":0,"BASE":18}

Device Name (Switch 1): Room1
Friendly Name 1 (Tasmota): Switch1
Friendly Name 2 (Tasmota): Switch2
Friendly Name 3 (Tasmota): Switch3
Friendly Name 1 (Tasmota): Switch4

Emulation: Hue Bridge multi device

The Queston:

Is it possible to get the Device to reset to custom factory settings, as shown in (a) and (b) above, when I issue a RESET Command or  press the RESET push button which Is hooked up to ground and RST Pin? Upon reset, I simply want the device to take the user to Tasmota access point where wifi settings can be configured and then it should already have above settings preloaded and activated.

My research so far indicates that I may have to compile my own Binary after making desired changes to user_config_override.h file. However, I am not sure how to enter above changes in user_config_override.h file. Can someone please help me with formatting above changes so that I can edit user_config_override.h file and compile custom bin file by using Gitpod or TasmoCompiler website. (Sorry, I am more of hardware person with minimal knowledge of coding.)

Thanks and Regards,

Navi

**Comment by sfromis** [comment]
All gpio settings as template in user_config_override.h
`#define MODULE USER_MODULE`
`#define USER_TEMPLATE "{\"NAME\":\"My template\",\"GPIO\":[160,1,256,1,1,1,1,1,1,1,1,1,1,1],\"FLAG\":0,\"BASE\":18}"`

**Reply by NsinghP** [reply]
Thanks for your reply, I did manage to open up _MY_USER_CONFIG_H_  and found the #define MODULE USER_MODULE field. A bit confusing at first. By the way what does #define FALLBACK_MODULE  do?

**Reply by sfromis** [reply]
You should make changes in user_config_override.h instead of my_user_config.h. FALLBACK_MODULE is meant to be a "more basic" config to revert to in case the "real" config cause boot failure. Getting a device locked in a boot loop is not fun, therefore it is best to have a "very minimal" fallback config.

**Comment by barbudor** [comment]
Hi
There are quite some settings that you can set into your `user_config_override.h` file to hard-code some default parameters in the binary.
You can check in `my_user_config.h` many examples of possible settings that you can copy and paste into `user_config_override.h` (do not modify my_user_config.h).

For example, setting the Wifi credentials:
```
#define STA_SSID1         "yourwifi"             // [Ssid1] Wifi SSID
#define STA_PASS1         "yourpassword"     // [Password1] Wifi password
```

Default module already answered by Sfromis

DeviceName cannot be set but will automatically take the value of FriendlyName1
```
  #define FRIENDLY_NAME         "YourFriendlyName"
```
FriendlyName2 to 4 cannot be set

There are similar settings to enable HUE Emulation

**Reply by NsinghP** [reply]
Thanks for your reply. I was able to find #define EMULATION              EMUL_NONE         // [Emulation] Select Belkin WeMo (single relay/light) or Hue Bridge emulation (multi relay/light) (EMUL_NONE, EMUL_WEMO or EMUL_HUE) in MY_USER_CONFIG_H. I think this the line I need to edit to EMUL HUE and then add to user_config_override.h

**Reply by NsinghP** [reply]
Ok so from all the helpful responses I have figured out that I need to make following Changes, please help me out by reviewing correctness of syntax and pointing out any errors in my assumption:

(1) On (platformio_override_sample.ini) activate line#L61 which is ( ;board_build.ldscript = eagle.flash.4m2m.ld ) to use autoexec.bat method

(2) In (user_config_override.h)  add from Line 53 of (_MY_USER_CONFIG_H_)
//#define MODULE                 MYHOME      // [Module] Select default module from tasmota_template.h

(3) In (user_config_override.h)  add from Line 56 of (_MY_USER_CONFIG_H_)
#define USER_TEMPLATE "{\"NAME\":\"MYHOME\",\"GPIO\":[258,259,1,163,256,257,1,1,162,161,160,1,1,1],\"FLAG\":0,\"BASE\":18}" 

(5) In (user_config_override.h) add from Line 172 of (_MY_USER_CONFIG_H_)
#define EMULATION              EMUL_HUE         // [Emulation] Select Belkin WeMo (single relay/light) or Hue Bridge emulation (multi relay/light) (EMUL_NONE, EMUL_WEMO or EMUL_HUE)

(6) In (user_config_override.h) add from Line 171 of (_MY_USER_CONFIG_H_)
#define FRIENDLY_NAME          "ROOM1"         // [FriendlyName] Friendlyname up to 32 characters used by webpages and Alexa

(7) (NOT SURE ABOUT THIS ONE) In (user_config_override.h) add from Line 55 of (_MY_USER_CONFIG_H_)

#define FALLBACK_MODULE        MYHOME      // [Module2] Select default module on fast reboot where USER_MODULE is user template

**Reply by barbudor** [reply]
@NsinghP 

(1) strictly speaking this is to enable File System. Once you have File System you can store autoexec.bat
You also need to rename `platformio_override_sample.ini` to `platformio_override.ini`
And add to your `user_config_override.h`
```
   #define USE_UFILESYS
      #define GUI_TRASH_FILE
```

(2) is wrong
`#define MODULE USER_MODULE`
your not meant to replace USER_MODULE by MYHOME. USER_MODULE is a specific value which instruct Tasmota to apply USER_TEMPLATE

(7) FALLBACK_MODULE shoud preferably be
`#define FALLBACK_MODULE WEMOS`
or
`#define FALLBACK_MODULE SONOFF_BASIC`
This instruct Tasmota to revert the configuration to a simple well know configuration in case of boot loop

**Reply by NsinghP** [reply]
All Right, Thanks for your help. Please review following syntax and assumptions for any errors:

(1.a) On (platformio_override_sample.ini) activate line#L61 wich is (;board_build.ldscript = eagle.flash.4m2m.ld) to use autoexec.bat method
(1.b) rename platformio_override_sample.ini to platformio_override.ini
(1.c) In (user_config_override.h)  add  #define USE_UFILESYS
(1.d) In (user_config_override.h) add   #define GUI_TRASH_FILE

(2) In (user_config_override.h)  add from Line 53 of (_MY_USER_CONFIG_H_)
//#define MODULE                 USER_MODULE      // [Module] Select default module from tasmota_template.h

(3) In (user_config_override.h)  add from Line 56 of (_MY_USER_CONFIG_H_)
#define USER_TEMPLATE "{\"NAME\":\"MYHOME\",\"GPIO\":[258,259,1,163,256,257,1,1,162,161,160,1,1,1],\"FLAG\":0,\"BASE\":18}" 

(5) In (user_config_override.h) add from Line 172 of (_MY_USER_CONFIG_H_)
#define EMULATION              EMUL_HUE         // [Emulation] Select Belkin WeMo (single relay/light) or Hue Bridge emulation (multi relay/light) (EMUL_NONE, EMUL_WEMO or EMUL_HUE)

(6) In (user_config_override.h) add from Line 171 of (_MY_USER_CONFIG_H_)
#define FRIENDLY_NAME          "ROOM1"         // [FriendlyName] Friendlyname up to 32 characters used by webpages and Alexa

(7) (No Need as already set to SONOFF_BASIC) In (user_config_override.h) add from Line 55 of (_MY_USER_CONFIG_H_)

#define FALLBACK_MODULE        SONOFF_BASIC      // [Module2] Select default module on fast reboot where USER_MODULE is user template

**Reply by barbudor** [reply]
(2) is still wrong. You left the // commenting the line and making it useless

(3) Each `"` inside the JSON string must be escaped by a `\` like in sfromis example:
`#define USER_TEMPLATE "{\"NAME\":\"MYHOME\",\"GPIO\":[258,259,1,163,256,257,1,1,162,161,160,1,1,1],\"FLAG\":0,\"BASE\":18}"`

**Reply by NsinghP** [reply]
Thanks, Corrected that error. Next I have to figure out how to use PlatformIO on Visual Studio Code and compile the bin file.

**Reply by barbudor** [reply]
https://tasmota.github.io/docs/Visual-Studio-Code/

**Reply by rapi3** [reply]
> ... Next I have to figure out how to use PlatformIO on Visual Studio Code and compile the bin file.

try to use this - easy and powerful without installing anything on your computer.
https://github.com/benzino77/tasmocompiler

**Reply by NsinghP** [reply]
Hello Rapi or Barbudor or anyone with knowledge,

I used tasmocompiler to successfully build a bin file with settings I wanted. However, I was not able to do the part that involves platformio_override_sample.ini file! which are as follows:

 (1.a) On (platformio_override_sample.ini) activate line#L61 which is (;board_build.ldscript = eagle.flash.4m2m.ld) to use autoexec.bat method
(1.b) rename platformio_override_sample.ini to platformio_override.ini

Is it even possible to do above two steps in online tasmocompiler or do I have to use Visual Studio Code method for that?

I also have a question on reset button/reset command function but I will post it as a new discussion.

Regards,

Navi

**Reply by rapi3** [reply]
Hello,
Latest Tasmocompiler claim that is doing that part auto if you select wemos device, so all you have to do is to insert all your custom  definition including:
#define USE_UFILESYS
#define GUI_TRASH_FILE
in the last window and select development and flash.
If that is not working let us.

p.s.
have a look before starting at all this discussion about this problem and freq regression.
https://github.com/benzino77/tasmocompiler/discussions/121

edit:
I just tested latest Tasmocompiler and it work ok compiling with UFS 2Mb if you add define lines.
Just to be sure I forked and added Freq 160Mhz so you can try the fork if you want also 160Mhz:
https://gitpod.io/#https://github.com/rapi3/tasmocompiler-new

**Reply by Jason2866** [reply]
The actual Tasmocompiler does not support using a different linker script.
Afaik benzino77 is working on a new version which will support that.

**Reply by NsinghP** [reply]
> 
> 

> 
> edit:
> I just tested latest Tasmocompiler and it work ok compiling with UFS 2Mb if you add define lines.
> Just to be sure I forked and added Freq 160Mhz so you can try the fork if you want also 160Mhz:
> https://gitpod.io/#https://github.com/rapi3/tasmocompiler-new

Hi rapi3, I tried your link. I was able to successfully compile bin file with my module configuration settings. But the file management feature did not get activated. I wonder what I did wrong? here are the lines i entered:

#define USE_UFILESYS
#define GUI_TRASH_FILE
#define MODULE            USER_MODULE
#define USER_TEMPLATE "{\"NAME\":\"MYROOM\",\"GPIO\":[258,259,32,163,256,257,1,1,162,161,160,1,1,1],\"FLAG\":0,\"BASE\":18}"
#define EMULATION         EMUL_HUE
#define FRIENDLY_NAME          "First Room Switch"

Regards,

Navi

**Reply by rapi3** [reply]
Interesting, did you select wemos module ?

if you look in the file:
https://github.com/rapi3/tasmocompiler-new/blob/master/src/components/AppStepper/FeaturesStep/AvailableBoards.js
you can see that only Wemos/NodeMCU UFS and SonOff Zigbee Bridge have UFS activated

**Reply by NsinghP** [reply]
> Interesting, did you select wemos module ?

Ok it worked when I selected Wemos Module. Now I have File Management System active. !!! Thanks. But there is one problem. Philips Hue emulation is gone!!! I guess wemos module does not have that?

**Reply by rapi3** [reply]
what device do you have and what options did you select first time when it was ok ?
you can try again and save your files used after compile and use it also for wemos what you need or look in this file and get it from there: https://github.com/arendst/Tasmota/blob/development/tasmota/my_user_config.h

// -- HTTP ----------------------------------------
#define USE_WEBSERVER                            // Enable web server and Wifi Manager (+66k code, +8k mem)
  #define WEB_PORT             80                // Web server Port for User and Admin mode
  #define WEB_USERNAME         "admin"           // Web server Admin mode user name
//  #define USE_JAVASCRIPT_ES6                     // Enable ECMAScript6 syntax using less JavaScript code bytes (fails on IE11)
  #define USE_ENHANCED_GUI_WIFI_SCAN             // Enable wifi scan output with BSSID (+0k5 code)
//  #define USE_WEBSEND_RESPONSE                   // Enable command WebSend response message (+1k code)
  #define USE_EMULATION_HUE                      // Enable Hue Bridge emulation for Alexa (+14k code, +2k mem common)
  #define USE_EMULATION_WEMO                     // Enable Belkin WeMo emulation for Alexa (+6k code, +2k mem common)

edit
read the doc for tasmocompiler....

**Reply by NsinghP** [reply]
OK,  #define USE_EMULATION_HUE , seemed to work as I saw the enable hue option under configure other. But after reset 1 command, the device seems to be bricked. I will serial flash it maybe tomorrow. I have a device that i made using ESP8266-12F chip.
![ESP8266 device](https://user-images.githubusercontent.com/69470381/114233291-b3790c00-999a-11eb-9a9b-ed51572e613e.jpeg)

**Reply by rapi3** [reply]
I updated the fork as wemos it was missing display and zigbee.
You just have to be sure to select all you need or just insert it in user define box before you compile it.

**Reply by NsinghP** [reply]
As I could see a "Hidden Network" in my computer's wifi scan, I had a hunch that device was not bricked, and that it was just not showing the Tasmota access point. I was able to reset my device back to factory settings by following the procedure for "Fast Power Cycle Device Recovery" so there was no need for serial reflashing. The Button1 method did not work. Restored device seems to be working as desired. I am still not sure why the device crashed? I am attaching screen shots below:

![Screenshot (87)](https://user-images.githubusercontent.com/69470381/114258468-41c5b000-99e4-11eb-9e05-18f1479cf0e8.png)
![Screenshot (91)](https://user-images.githubusercontent.com/69470381/114301462-a9622500-9ae2-11eb-9289-763fff923a1a.png)



![Screenshot (86)](https://user-images.githubusercontent.com/69470381/114258472-47bb9100-99e4-11eb-97bb-34a24b517744.png)

**Reply by NsinghP** [reply]
> DeviceName cannot be set but will automatically take the value of FriendlyName1
> 
> ```
>   #define FRIENDLY_NAME         "YourFriendlyName"
> ```

Hello barbadour, or anyone with knowledge on the issue,

Can Friendly_Name feature be modified easily? It is rather inconvenient to not be able to set a device name that is not dependent on Friendly Name 1. As you can see in screenshot above I have set the Friendly Name 1 to "First Room Switch" as a result Device name has automatically become "First Room Switch". I would rather have the Device name as "First Room" and Friendly name 1 as "First Room Switch 1" but I can't do that because it will change Device name to ""First Room Switch 1". 

Is there any workaround?
Can this feature be requested? 
Is there a reason, that I don't understand, which prevents implementation of this feature?

Just curious, because it will be nice to be able to name the device independently of Switch Names.

Regards,

Navi

**Comment by Jason2866** [comment, marked as answer]
Since a ESP8266-12F has a 4MB flash you can build a variant with a filesystem and store settings
you want to use in `autoexec.bat`. All commands can be used which do NOT need a restart of Tasmota.
The file partition does NOT get erased by all kind of reset commands. Only a flash erase will wipe.

**Reply by rapi3** [reply]
This is interesting but is it somewhere documented the filesystem how to create & use ?
thank you

**Reply by Jason2866** [reply]
Yes, there is documentation work to do. Maybe you?
Activate this line https://github.com/arendst/Tasmota/blob/development/platformio_override_sample.ini#L61 
and you will find a new entry in Tasmota web GUI `Manage Filesystem`. You can upload your `autoexec.bat`
![Screenshot_20210406-162414](https://user-images.githubusercontent.com/24528715/113726543-9d4a2200-96f4-11eb-81ac-071505bfacd3.png)
Every command in a extra line in autoexec.bat will be executed sequentially

**Reply by NsinghP** [reply]
Thank you for your reply. If I use autoexec.bat method then I don't need to compile a fresh binary? Is there a sample of autoexec.bat file that can be used? 

Lin#61 is ;board_build.ldscript        = eagle.flash.4m2m.ld 
Should I activate this line?

**Reply by rapi3** [reply]
No problem I can write the doc only if I understand properly how to use this, after/if I will succeed to compile/activate.

**Reply by rapi3** [reply]
Are you sure that is all I need to change to have File system active in latest development ?
Nothing else to define ?
Because I still can't see it after I compiled... :(

**Reply by barbudor** [reply]
You need to rename `platformio_override_sample.ini` as `platformio_override.ini` for the change to be taken into account
And you need those 2 lines in your `user_config_override.h`:
```
    #define USE_UFILESYS
      #define GUI_TRASH_FILE
```

**Reply by barbudor** [reply]
@NsinghP Any command that you would type in the console works
Like creating and enabling rules, ...
Avoid any command that will make the device to reboot like changing Wifi setting, MQTT settings, templates & GPIO, ...
Otherwise it will reboot continuously (in fact 3 or 4 times then autoexec will be disabled after boot loop detection)

**Reply by rapi3** [reply]
> You need to rename `platformio_override_sample.ini` as `platformio_override.ini` for the change to be taken into account
> And you need those 2 lines in your `user_config_override.h`:
> 
> ```
>     #define USE_UFILESYS
>       #define GUI_TRASH_FILE
> ```

thank you for your complete information.
Of course I was missing the 2 define lines.
I can start experimenting further now.
![Screenshot_2021-04-06_19-14-54](https://user-images.githubusercontent.com/18531150/113751792-929b8700-970c-11eb-835a-a1bf51e9d047.png)
